renku/service/controllers/api/mixins.py

```diff
@@ -203,27 +222,43 @@ def execute_op(self):
 
     def reset_local_repo(self, project):
         """Reset the local repo to be up to date with the remote."""
-        repo = Repo(project.abs_path)
-        origin = None
-        tracking_branch = repo.active_branch.tracking_branch()
-        if tracking_branch:
-            origin = repo.remotes[tracking_branch.remote_name]
-        elif repo.remotes and len(repo.remotes) == 1:
-            origin = repo.remotes[0]
-
-        if origin and not self.migrate_project:
-            origin.fetch(repo.active_branch)
-            repo.git.reset("--hard", f"{origin}/{repo.active_branch.name}")
-
-        elif origin and self.migrate_project:
-            try:
-                # NOTE: It could happen that repository is already un-shallowed,
-                # in this case we don't want to leak git exception, but still want to fetch.
-                origin.fetch(repo.active_branch, unshallow=True)
-            except GitCommandError:
-                origin.fetch(repo.active_branch)
 
-            repo.git.reset("--hard", f"{origin}/{repo.active_branch}")
+        # NOTE: Only do a fetch every >30s to get eventual consistency but not slow things down too much
+        if project.fetch_age < PROJECT_FETCH_TIME:
+            return
+
+        lock = project.write_lock()
+
+        if self.skip_lock:
+            lock = contextlib.suppress()
+        try:
+            with lock:
```

Comment:
Should we again check if `project.fetch_age < PROJECT_FETCH_TIME` again here and do nothing if it's the case? This makes sense if multiple operations are waiting for the lock to do the reset.